docs/src/content/docs/intermediate/configure-php-xdebug.md
---
title: "Configure PHP Xdebug"
---

# Configure PHP Xdebug

This section explains in depth how to enable and use PHP Xdebug with the
Devilbox.


## Introduction

In order to have a working Xdebug, you need to ensure two things:

1.  **PHP Xdebug** must be configured and enabled in PHP itself
2.  Your **IDE/editor** must be configured and requires a way talk to
    PHP

Configuring PHP Xdebug will slightly differ when configuring it for a
dockerized environment. This is due to the fact that Docker versions on
different host os have varying implementations of how they connect back
to the host.

Most IDE or editors will also require different configurations for how
they talk to PHP Xdebug. This is at least most likely the case for
`xdebug.idekey`.

<div class="seealso">

`configure-php-xdebug-options`

</div>

## Configure Xdebug

### Docker on Linux

Docker on Linux allows Xdebug to automatically connect back to the host
system without the need of an explicit IP address.

<div class="toctree" glob="" maxdepth="1" hidden="">

/intermediate/configure-php-xdebug/linux/\*

</div>

<div class="seealso">

- `configure-php-xdebug-lin-atom`
- `configure-php-xdebug-lin-phpstorm`
- `configure-php-xdebug-lin-sublime`
- `configure-php-xdebug-lin-vscode`

</div>

### Docker on MacOS

<div class="toctree" glob="" maxdepth="1" hidden="">

/intermediate/configure-php-xdebug/macos/\*

</div>

<div class="seealso">

- `configure-php-xdebug-mac-atom`
- `configure-php-xdebug-mac-phpstorm`
- `configure-php-xdebug-mac-sublime`
- `configure-php-xdebug-mac-vscode`

</div>

### Docker on Windows

<div class="toctree" glob="" maxdepth="1" hidden="">

/intermediate/configure-php-xdebug/windows/\*

</div>

<div class="seealso">

- `configure-php-xdebug-win-atom`
- `configure-php-xdebug-win-phpstorm`
- `configure-php-xdebug-win-sublime`
- `configure-php-xdebug-win-vscode`

</div>

### Docker Toolbox

Docker Toolbox configuration is equal, no matter if it is started on
MacOS or Windows, as both use a Linux system inside VirtualBox.

<div class="toctree" glob="" maxdepth="1" hidden="">

/intermediate/configure-php-xdebug/toolbox/\*

</div>

<div class="seealso">

- `configure-php-xdebug-toolbox-atom`
- `configure-php-xdebug-toolbox-phpstorm`
- `configure-php-xdebug-toolbox-sublime`
- `configure-php-xdebug-toolbox-vscode`

</div>


docs/src/content/docs/_includes/figures/https/chrome-settings.md
---
title: "Click on ``Advanced``"
sidebar:
  hidden: true
---

<figure>
<img src="/_includes/figures/https/chrome-settings.png"
alt="Click on Advanced" />
<figcaption aria-hidden="true">Click on
<code>Advanced</code></figcaption>
</figure>


CHANGELOG.md
# Changelog

Make sure to have a look at [UPDATING.md](https://github.com/cytopia/devilbox/blob/master/UPDATING.md) to see any required steps for updating major or minor versions.


## Unreleased


## Release v3.0.1 (2025-07-17)

### Added
- Officially use `buggregator` service for easier development workflow
- New sub-command `cloud-cli` (Magento Cloud CLI) for `dvl.sh`
- Support `mailpit` as drop-in replacement for `mailhog` (port conflicts with `buggregator`)

### Changed
- Updated PHP image version to `0.160`
- Updated softwares of image to latest versions
- Resolve some issues with `msmtp` package
- Drop support for `mailhog` officially


## Release v3.0.0 (2025-03-27)

### Added
- Support PHP 8.4
- Simplified the initialize webapp process of `dvl.sh`
- Wrapping all PHP related commands with a detection to check and use PHP version
- New sub-command `ece-tools` for `dvl.sh`
- New sub-command `cloud-patches` for `dvl.sh`
- New sub-command `update-docroot` for `dvl.sh`
- New sub-command `sync-env` for `dvl.sh`

### Changed
- Updated PHP image version to `0.155`
- Updated softwares of image to latest versions


## Release v3.0.0-beta-0.7 (2024-11-05)

## Added
- New sub-command `magerun` for `dvl.sh`
- New sub-command `db:import` (or `db-import`) for `dvl.sh`
- Add a new way to initialize project for `dvl init` (via `.devilbox.yaml`)

## Changed
- Removed unsupported and EOL PHP versions out of Devilbox
- Update PHP image version to `0.154` (remove EOL versions, PHP EOL versions still available in `0.153` and previous)


## Release v3.0.0-beta-0.6 (2024-06-01)

### Added
- New sub-command `magento` for `dvl.sh`
- New sub-command `composer` for `dvl.sh`

### Changed
- Re-added and upgrade `mhsendmail` tool
- Update PHP image version to `0.153`
- Change default flavour for PHP image in order to use `mailhog`


## Release v3.0.0-beta-0.5 (2024-06-18)

### Added
- Support PHP 8.3
- New shell script `dvl.sh` which support initialize projects and all daily stuff of Devilbox

### Changed
- Replace `mhsendmail` with `msmtp`
- Remove unsupported PHP version 5.2 -> 5.5
- Default TLD_SUFFIX now changed to `lvh.me`


## Release v3.0.0-beta-0.4 (2022-01-30)

### Fixed
- Ignore directories starting with a `.` in `/shared/httpd` (invalid server names)
- Fixed: `Creation of dynamic property Mail_Mbox::$_file is deprecated` (PHP 8.2)
- Fixed: Xdebug default configuration files in cfg/php-ini-X.Y/devilbox-php.ini-xdebug [#866](https://github.com/cytopia/devilbox/issues/866)
- Fixed: Show httpd access logs via docker logs when `DOCKER_LOGS=1` instead of piping it to supervisord [#956](https://github.com/cytopia/devilbox/issues/956)

### Added
- Intranet: Allow to reload PHP-FPM to reflect `php.ini` or `php-fpm.conf` changes on-the-fly without restarting docker-compose

### Changed
- Updated httpd images
- Updated php-fpm images


## Release v3.0.0-beta-0.3 (2022-01-02)

This release provides the `dvl.to` domain to be used with `TLD_SUFFIX` (set to default), which eliminates the need to set any entries in `/etc/hosts`, as all of its subdomain will point to `127.0.0.1` via official DNS. Domain has been acquired thanks to awesome sponsors!

### Fixed
- Intranet: mail.php fixed deprecation warnings [#798](https://github.com/cytopia/devilbox/issues/798)
- Added `host.docker.internal` to extra_hosts to be able to connect to the host system [#919](https://github.com/cytopia/devilbox/issues/919)

### Changed
- Use `dvl.to` as default `TLD_SUFFIX` (it always poits to `127.0.0.1` removing the need to create `/etc/hosts` entries)

### Added
- Intranet: vhost overview shows listening ports
- Intranet: vhost overview now has modals to show httpd and vhost-gen configs
- Docs: Show available tools per version in README.md
- Added `xhprof` PHP extension


## Release v3.0.0-beta-0.2 (2022-12-27)

The Backend configuration now supports websockets as well:

file: `/shared/httpd/<project>/.devilbox/backend.cfg`
```bash
# PHP-FPM backend
conf:phpfpm:tcp:php80:9000

# HTTP Reverse Proxy backend
conf:rproxy:http:172.16.238.10:3000

# HTTPS Reverse Proxy backend
conf:rproxy:https:172.16.238.10:3000

# Websocket Reverse Proxy backend
conf:rproxy:ws:172.16.238.10:3000

# SSL Websocket Reverse Proxy backend
conf:rproxy:wss:172.16.238.10:3000
```

Once you're done with `backend.cfg` changes, head over to the Intranet C&C page (http://localhost/cnc.php) and Reload `watcherd`.


### Fixed
- Intranet: vhost overview: allow HTTP 426 to succeed in vhost page (websocket projects)
- Intranet: vhost overview: Reverse Proxy or Websocket backends do not require a `htdocs/` dir for healthcheck
- Fixed reverse proxy template generation for Apache 2.2 and Apache 2.4 [vhost-gen #51](https://github.com/devilbox/vhost-gen/pull/51)
- Fixed Nginx hash bucket size length to allow long hostnames

### Added
- Reverse Proxy automation for websocket projects (`ws://<host>:<port>` or `wss:<host>:<port>`) (Does not work with Apache 2.2)
- Added tool `wscat` to be able to test websocket connections
- Intranet: show `wscat` version
- Intranet: vhost overview now also shows websocket projects

### Changed
- Do not mount any startup/autostart script directories for multi-php compose as they do not contain tools
- Updated vhost-gen templates in `cfg/vhost-gen` (replace your project templates with new ones)


## Release v3.0.0-beta-0.1 (2022-12-24) 🎅🎄🎁

This is a beta release, using a completely rewritten set of HTTPD server, which allow easy reverse Proxy integration and different PHP versions per project:

* https://github.com/devilbox/docker-nginx-stable/pull/55
* https://github.com/devilbox/docker-nginx-mainline/pull/57
* https://github.com/devilbox/docker-apache-2.2/pull/53
* https://github.com/devilbox/docker-apache-2.4/pull/54

Once it has been tested by the community, and potential errors have been addressed, a new major version will be released.


.devilbox/www/config.php
<?PHP
// Measure time
$TIME_START = microtime(true);

// Start session
session_start();

// Turn on all PHP errors
error_reporting(-1);


// Shorten DNS timeouts for gethostbyname in case DNS server is down
putenv('RES_OPTIONS=retrans:1 retry:1 timeout:1 attempts:1');


$DEVILBOX_VERSION = 'v3.0.1';
$DEVILBOX_DATE = '2025-07-17';
$DEVILBOX_API_PAGE = 'devilbox-api/status.json';

//
// Set Directories
//
$CONF_DIR	= dirname(__FILE__);
$LIB_DIR	= $CONF_DIR . DIRECTORY_SEPARATOR . 'include' . DIRECTORY_SEPARATOR .'lib';
$VEN_DIR	= $CONF_DIR . DIRECTORY_SEPARATOR . 'include' . DIRECTORY_SEPARATOR .'vendor';
$LOG_DIR	= dirname(dirname($CONF_DIR)) . DIRECTORY_SEPARATOR . 'log' . DIRECTORY_SEPARATOR . 'devilbox';


//
// Load Base classes
//
require $LIB_DIR . DIRECTORY_SEPARATOR . 'container' . DIRECTORY_SEPARATOR .'BaseClass.php';
require $LIB_DIR . DIRECTORY_SEPARATOR . 'container' . DIRECTORY_SEPARATOR .'BaseInterface.php';



//
// Set Docker addresses
//
$DNS_HOST_NAME		= 'bind';
$PHP_HOST_NAME		= 'php';
$HTTPD_HOST_NAME	= 'httpd';
$MYSQL_HOST_NAME	= 'mysql';
$PGSQL_HOST_NAME	= 'pgsql';
$REDIS_HOST_NAME	= 'redis';
$MEMCD_HOST_NAME	= 'memcd';
$MONGO_HOST_NAME	= 'mongo';


//
// Lazy Container Loader
//
function loadFile($class, $base_path) {
	static $_LOADED_FILE;

	if (isset($_LOADED_FILE[$class])) {
		return;
	}

	require $base_path . DIRECTORY_SEPARATOR . $class . '.php';
	$_LOADED_FILE[$class] = true;
	return;
}
function loadClass($class) {

	static $_LOADED_LIBS;

	if (isset($_LOADED_LIBS[$class])) {
		return $_LOADED_LIBS[$class];
	} else {

		$lib_dir = $GLOBALS['LIB_DIR'];
		$cnt_dir = $GLOBALS['LIB_DIR'] . DIRECTORY_SEPARATOR . 'container';

		switch($class) {
			//
			// Lib Classes
			//
			case 'Logger':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \devilbox\Logger::getInstance();
				break;

			case 'Html':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \devilbox\Html::getInstance();
				break;

			case 'Helper':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \devilbox\Helper::getInstance();
				break;

			//
			// Docker Container Classes
			//
			case 'Php':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \devilbox\Php::getInstance($GLOBALS['PHP_HOST_NAME']);
				break;

			case 'Dns':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \devilbox\Dns::getInstance($GLOBALS['DNS_HOST_NAME']);
				break;

			case 'Httpd':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \devilbox\Httpd::getInstance($GLOBALS['HTTPD_HOST_NAME']);
				break;

			case 'Mysql':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \devilbox\Mysql::getInstance($GLOBALS['MYSQL_HOST_NAME'], array(
					'user' => 'root',
					'pass' => loadClass('Helper')->getEnv('MYSQL_ROOT_PASSWORD')
				));
				break;

			case 'Pgsql':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \devilbox\Pgsql::getInstance($GLOBALS['PGSQL_HOST_NAME'], array(
					'user' => loadClass('Helper')->getEnv('PGSQL_ROOT_USER'),
					'pass' => loadClass('Helper')->getEnv('PGSQL_ROOT_PASSWORD'),
					'db' => 'postgres'
				));
				break;

			case 'Redis':
				// Check if redis is using a password
				$REDIS_ROOT_PASSWORD = '';

				$_REDIS_ARGS = loadClass('Helper')->getEnv('REDIS_ARGS');

				/*
				 * This pattern will match optional quoted string, 'my password' or "my password"
				 * or if there aren't any quotes, it will match up until the next space.
				 */
				$_REDIS_PASS = array();
				preg_match_all('/--requirepass\s+("|\')?(?(1)(.*)|([^\s]*))(?(1)\1|)/', $_REDIS_ARGS, $_REDIS_PASS, PREG_SET_ORDER);

				if (! empty($_REDIS_PASS)) {
					/*
					 * In case the option is specified multiple times, use the last effective one.
					 *
					 * preg_match_all returns a multi-dimensional array, the first level array is in order of which was matched first,
					 * and the password string is either matched in group 2 or group 3 which is always the end of the sub-array.
					 */
					$_REDIS_PASS_ROW = end($_REDIS_PASS);
					$_REDIS_PASS = end($_REDIS_PASS_ROW);

					if (strlen($_REDIS_PASS) > 0) {
						$REDIS_ROOT_PASSWORD = $_REDIS_PASS;
					}
				}

				loadFile($class, $cnt_dir);
				if ($REDIS_ROOT_PASSWORD == '') {
					$_LOADED_LIBS[$class] = \devilbox\Redis::getInstance($GLOBALS['REDIS_HOST_NAME']);
				} else {
					$_LOADED_LIBS[$class] = \devilbox\Redis::getInstance($GLOBALS['REDIS_HOST_NAME'], array(
						'pass' => $REDIS_ROOT_PASSWORD,
					));
				}
				break;

			case 'Memcd':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \devilbox\Memcd::getInstance($GLOBALS['MEMCD_HOST_NAME']);
				break;

			case 'Mongo':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \devilbox\Mongo::getInstance($GLOBALS['MONGO_HOST_NAME']);
				break;

			// Get optional docker classes
			default:
				// Unknown class
				exit('Class does not exist: '.$class);
		}
		return $_LOADED_LIBS[$class];
	}
}
